### PR TITLE
Add a custom cronjob functionality

### LINKED ISSUE #67: Add a custom cronjob entry (not daily, weekly, monthly)
# 💡 Summary #

I want to add a custom cronjob timer to assure when the job start

## Motivation and context ##

We already create a cronjob, we just don't handle when it starts. It's useful because on Debian systems, daily start at 6:47 ( as stated here : https://wiki.debian.org/cron ). It's too late and without randomization, which put too many stress in our virtualization infrastructure.

## Implementation notes ##

By a variable, we set a custom timer to launch our already set cronjob.

## Acceptance criteria ##

- [  ] We can add the cronjob timer with a simple variable

### PR DESCRIPTION
**Pull request by ClementJ35**
# Custom cronjob #

## 🗣 Description ##

This PR adds a custom cronjob functionality, handled by an Ansible variables dictionary. This resolves #67.

## 💭 Motivation and context ##

Especially on Debian, cronjobs are launched at the same time when we use cron keyword as daily. This is hard to sustain with an entire virtualization platform on which every scan starts at the exact same time.

## 🧪 Testing ##

I tested on our virtual RedHat and Debian servers.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

### CODE REVIEW

**Comment by ClementJ35** [comment]
Beware : This PR change the cron job behavior which can lead in two cron jobs both actives. The old one and the new one.
The old way may lead to the same problem when changing from daily to weekly.
Now, it's always the same file in /etc/cron.d/virus_scan that is modified.

I don't know if the cleaning facility of the old files must be handled here.

**Comment by ClementJ35** [comment]
Hi,
Didn't have much news about this PR.
What do you think of it @jsf9k @mcdonnnj ?
And what about cleaning /etc/cron.*/virus_scan old files ?

IMO it should not be integrated here as it's a one shot, but it could be another little playbook related to this project.

File: defaults/main.yml [hunk 1]
```diff
@@ -1,5 +1,17 @@
+---
+# Frequency of cronjob, must be one of "custom" or a special_time handled by ansible.builtin.cron
```
Comment (jsf9k) [inline defaults/main.yml]:
```suggestion
# Frequency of clamav cronjob, must be one of "custom" or a special_time handled by ansible.builtin.cron
```

File: defaults/main.yml [hunk 2]
```diff
@@ -1,5 +1,17 @@
...
 clamav_cron_frequency: weekly
 
+# If custom frequency, the role uses this variable to install the cronjob.
+# This dict doesn't merge, every line needs to be valuated:
```
Comment (jsf9k) [inline defaults/main.yml]:
```suggestion
# If clamav cronjob frequency is custom, the role uses this variable to install the cronjob.
# This dict doesn't merge, every line needs to be present.
```

File: templates/virus_scan.sh.j2 [hunk 3]
```diff
@@ -1,5 +1,6 @@
 #!/bin/bash
 # Managed by ansible role clamscan
+# Beware : this script will be run by root
```
Comment (jsf9k) [inline templates/virus_scan.sh.j2]:
```suggestion
# Beware:  This script will be run by root.
```

File: tasks/main.yml [hunk 4]
```diff
@@ -16,12 +16,34 @@
...
   ansible.builtin.template:
-    dest: /etc/cron.{{ clamav_cron_frequency }}/virus_scan
+    dest: /usr/local/share/virus_scan.sh
     mode: 0755
     src: virus_scan.sh.j2
 
+- name: "Install virus_scan cron job as {{ clamav_cron_frequency }}"
+  ansible.builtin.cron:
+    cron_file: virus_scan
+    job: "/usr/local/share/virus_scan.sh"
```
Comment (jsf9k) [inline tasks/main.yml]: The quotes are unnecessary here:
```suggestion
    job: /usr/local/share/virus_scan.sh
```

File: molecule/default/tests/test_files_and_dirs.py [hunk 5]
```diff
@@ -16,7 +16,9 @@
     "path",
     [
         # The virus scan cron job
-        "/etc/cron.weekly/virus_scan",
+        "/etc/cron.d/virus_scan",
+        # The virus scan shell
```
Comment (jsf9k) [inline molecule/default/tests/test_files_and_dirs.py]:
```suggestion
        # The virus scan shell script
```

File: tasks/main.yml [hunk 6]
```diff
@@ -16,12 +16,34 @@
...
 
-- name: Install virus_scan cron job
+- name: Install shell script to be run by cron
   ansible.builtin.template:
-    dest: /etc/cron.{{ clamav_cron_frequency }}/virus_scan
+    dest: /usr/local/share/virus_scan.sh
     mode: 0755
     src: virus_scan.sh.j2
 
+- name: "Install virus_scan cron job as {{ clamav_cron_frequency }}"
```
Comment (jsf9k) [inline tasks/main.yml]: Quotes are not necessary here:
```suggestion
- name: Install virus_scan cron job as {{ clamav_cron_frequency }}
```

**Review by jsf9k — CHANGES_REQUESTED** [review]
I'm sorry, this PR somehow fell off my radar.  I requested some minor cleanup changes, but overall this looks good to me.  Thank you for the contribution!

**Comment by jsf9k** [comment]
> Hi, Didn't have much news about this PR. What do you think of it @jsf9k @mcdonnnj ? And what about cleaning /etc/cron.*/virus_scan old files ?
> 
> IMO it should not be integrated here as it's a one shot, but it could be another little playbook related to this project.

Apologies, this PR somehow fell off my radar.  I don't think we need to worry about cleaning the old files, but if you want to create a separate playbook in an `extras` directory go ahead.

**Comment by ClementJ35** [comment]
I committed a little playbook to clean all cronjobs.
That's good to me.

File: tasks/main.yml [hunk 7]
```diff
@@ -16,12 +16,34 @@
...
+- name: Install virus_scan cron job as {{ clamav_cron_frequency }}
+  ansible.builtin.cron:
+    cron_file: virus_scan
+    job: /usr/local/share/virus_scan.sh
+    name: virus_scan
+    special_time: "{{ clamav_cron_frequency }}"
+    user: root
+  when: clamav_cron_frequency != "custom"
+
+- name: "Install custom virus_scan cron job"
```
Comment (jsf9k) [inline tasks/main.yml]: Remove unnecessary quotes:
```suggestion
- name: Install custom virus_scan cron job
```

File: defaults/main.yml [hunk 8]
```diff
@@ -1,5 +1,17 @@
+---
+# Frequency of clamav cronjob, must be one of "custom" or a special_time handled by ansible.builtin.cron
 clamav_cron_frequency: weekly
 
+# If clamav cronjob frequency is custom, the role uses this variable to install the cronjob.
```
Comment (jsf9k) [inline defaults/main.yml]:
```suggestion
# If clamav cronjob frequency is "custom", the role uses this variable to install the cronjob.
```

File: defaults/main.yml [hunk 9]
```diff
@@ -1,5 +1,17 @@
+---
+# Frequency of clamav cronjob, must be one of "custom" or a special_time handled by ansible.builtin.cron
```
Comment (jsf9k) [inline defaults/main.yml]:
```suggestion
# Frequency of clamav cronjob, must be "custom" or a special_time handled by ansible.builtin.cron
```

File: defaults/main.yml [hunk 10]
```diff
@@ -1,5 +1,17 @@
...
 clamav_cron_frequency: weekly
 
+# If clamav cronjob frequency is custom, the role uses this variable to install the cronjob.
+# This dict doesn't merge, every line needs to be present.
```
Comment (jsf9k) [inline defaults/main.yml]:
```suggestion
# This dict doesn't merge, so every line needs to be present.
```

**Review by jsf9k — CHANGES_REQUESTED** [review]
A few more minor language tweaks.

**Comment by jsf9k** [comment]
@ClementJ35 - When you accept someone's suggested changes, please change the commit message to something that describes the change.  Imagine someone who revisits these commits in the future; something like "Remove unnecessary quotes" will tell them much more about what changes the commit contains than "Apply suggestions from code review" or "Update defaults/main.yml".

**Review by jsf9k — APPROVED** [review]
Thank you for the contribution @ClementJ35!

File: README.md [hunk 11]
```diff
@@ -19,7 +19,8 @@ None.
 |----------|-------------|---------|----------|
 | clamav_clamd_configuration | A dictionary of values to set in the clamd configuration file. | `{}` | No |
 | clamav_configuration_backup | Whether or not to backup configuration files before changing. | `false` | No |
-| clamav_cron_frequency | The frequency of ClamAV scanning.  Must be one of: `hourly`, `daily`, `weekly`, or `monthly`. | `weekly` | No |
+| clamav_cron_frequency | The frequency of ClamAV scanning.  Must be `custom` or an ansible.builtin.cron [special_time](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/cron_module.html#parameter-special_time). | `weekly` | No |
+| clamav_cron_custom    | If frequency is set to custom a dictionnary define the timer. | `05:30` | No |
```
Comment (dav3r) [inline README.md]: This cleans up the description, fixes a typo, and updates the correct default value.
```suggestion
| clamav_cron_custom    | If frequency is set to `custom`, a dictionary to define the timer. | `{"day": "*", "job": "/usr/local/share/virus_scan.sh", "minute": "30", "month": "*", "hour": "5", "weekday": "*"}` | No |
```

**Review by dav3r — CHANGES_REQUESTED** [review]
I see one more thing that needs to be cleaned up here.

**Review by dav3r — APPROVED** [review]
👍 👍  Thanks for your contribution to this repo!

**Review by jsf9k — APPROVED** [review]

**Comment by jsf9k** [comment]
Thanks for yet another contribution @ClementJ35!

**Pull request merged**